File: cpp/49.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        vector<vector<string>> res;
        unordered_map<string, int> hash;
        for(int i=0; i<strs.size(); i++){
            string str = strs[i];
            sort(strs[i].begin(), strs[i].end());
            if(hash.find(strs[i]) != hash.end()){
                res[hash[strs[i]]].push_back(str);
            }else{
                res.push_back(vector<string>{str});
                hash[strs[i]] = res.size()-1;
            }
        }
        return res;
    }
};

File: golang/49/49.go
package main

import "sort"

func groupAnagrams(strs []string) [][]string {
	mm := make(map[string][]string)
	for _, v := range strs {
		str := SortStringByCharacter(v)
		mm[str] = append(mm[str], v)
	}

	var res [][]string
	for _, v := range mm {
		res = append(res, v)
	}
	return res
}

func SortStringByCharacter(s string) string {
	r := []rune(s)
	sort.Slice(r, func(i, j int) bool {
		return r[i] < r[j]
	})
	return string(r)
}
